api/src/test/java/org/openmrs/util/ConceptReferenceRangeUtilityTest.java

```diff
@@ -0,0 +1,292 @@
+	}
+
+	@Test
+	public void testGenderMatch_shouldReturnFalseIfGenderIsNull() {
+		String criteria = "$patient.getGender().equals('M')";
+
+		assertFalse(ConceptReferenceRangeUtility.evaluateCriteria(criteria, person));
+	}
+
+	@Test
+	public void testAgeAndGenderMatch_shouldReturnTrueIfAgeAndGenderMatch() {
+		calendar.add(Calendar.YEAR, -5);
+		person.setBirthdate(calendar.getTime());
+		person.setGender("M");
+		String criteria = "$patient.getAge() > 1 && $patient.getAge() < 10 && $patient.getGender().equals('M')";
+
+		assertTrue(ConceptReferenceRangeUtility.evaluateCriteria(criteria, person));
+	}
+
+	@Test
+	public void testAgeAndGenderMatch_shouldReturnFalseIfAgeDoesNotMatch() {
+		calendar.add(Calendar.YEAR, -11);
+		person.setBirthdate(calendar.getTime());
+		person.setGender("M");
+		String criteria = "$patient.getAge() > 1 && $patient.getAge() < 10 && $patient.getGender().equals('M')";
+
+		assertFalse(ConceptReferenceRangeUtility.evaluateCriteria(criteria, person));
+	}
+
+	@Test
+	public void testAgeAndGenderMatch_shouldReturnFalseIfGenderDoesNotMatch() {
+		calendar.add(Calendar.YEAR, -5);
+		person.setBirthdate(calendar.getTime());
+		person.setGender("F");
+		String criteria = "$patient.getAge() > 1 && $patient.getAge() < 10 && $patient.getGender().equals('M')";
+
+		assertFalse(ConceptReferenceRangeUtility.evaluateCriteria(criteria, person));
+	}
+
+	@Test
+	public void testAgeAndGenderMatch_shouldReturnFalseIfPersonIsNull() {
+		String criteria = "$patient.getAge() > 1 && $patient.getAge() < 10 && $patient.getGender().equals('M')";
+
+		try {
+			ConceptReferenceRangeUtility.evaluateCriteria(criteria, null);
+		} catch (ValidationException e) {
+			Assertions.assertEquals("Failed to validate with reason: patient is null", e.getMessage());
+		}
+	}
+
+	@Test
+	public void testAgeAndGenderMatch_shouldReturnFalseIfAgeOrGenderIsNull() {
+		String criteria = "$patient.getAge() > 1 && $patient.getAge() < 10 && $patient.getGender().equals('M')";
+		assertFalse(ConceptReferenceRangeUtility.evaluateCriteria(criteria, person));
+	}
+
+	@Test
+	public void testObsValueMatch_shouldReturnTrueIfValueTextMatch() {
+		person.setGender("F");
+		
+		Obs obs = buildObservation();
+		obs.setPerson(person);
+		obs.setValueText("PREGNANT");
+
+		Mockito.when(obsService.getLatestObsByConceptId("5089")).thenReturn(obs);
+
+		String criteria = "$patient.getGender().equals('F') && $fn.getLatestObsByConceptId('5089').getValueText().equals('PREGNANT')";
+
+		assertTrue(ConceptReferenceRangeUtility.evaluateCriteria(criteria, person));
+	}
+
+	@Test
+	public void testObsValueMatch_shouldReturnFalseIfValueMismatch() {
+		person.setGender("M");
+
+		String criteria = "$patient.getGender().equals('F') && $fn.getLatestObsByConceptId('5089') == true";
+
+		assertFalse(ConceptReferenceRangeUtility.evaluateCriteria(criteria, person));
+	}
+
+	@Test
+	public void testTimeOfDay_shouldReturnTrueIfTimeMatches() {
+		int hour = LocalTime.now().getHour();
```

Comment:
What happens when the above call is made at 10:59:59 and by the type the evaluation is done, the time has become 11:00 for cases when this test is run on a slow machine?

Reply:
This would be a very rare scenario 😄 .

The solution here would be to Mock LocalTime which require powermock .

<exclusions>
				<exclusion>
					<groupId>org.powermock</groupId>
					<artifactId>powermock-module-junit4</artifactId>
				</exclusion>
				<exclusion>
					<groupId>org.powermock</groupId>
					<artifactId>powermock-api-mockito2</artifactId>
				</exclusion>
			</exclusions>

Reply:
And when that happens the test will fail. Can we make the test less brittle?